packages/timeline-state-resolver-types/src/tricaster.ts

```diff
@@ -0,0 +1,307 @@
+}
+
+/**
+ * Convenience object for the keyers in the Main M/E
+ */
+export interface TimelineObjTriCasterDSK extends TimelineObjTriCasterBase {
+	content: {
+		deviceType: DeviceType.TRICASTER
+		type: TimelineContentTypeTriCaster.DSK
+
+		keyer: TriCasterKeyer
+	} & TimelineDatastoreReferencesContent
+}
+
+export function isTimelineObjTriCasterDSK(
+	timelineObject: TSRTimelineObjBase
+): timelineObject is TimelineObjTriCasterDSK {
+	return (timelineObject as TimelineObjTriCasterBase).content?.type === TimelineContentTypeTriCaster.DSK
+}
+
+export interface TriCasterInput {
+	videoActAsAlpha?: boolean
+	videoSource?: string
+}
+
+export interface TimelineObjTriCasterInput extends TimelineObjTriCasterBase {
+	content: {
+		deviceType: DeviceType.TRICASTER
+		type: TimelineContentTypeTriCaster.INPUT
+
+		input: TriCasterInput
+	} & TimelineDatastoreReferencesContent
+}
+
+export function isTimelineObjTriCasterInput(
+	timelineObject: TSRTimelineObjBase
+): timelineObject is TimelineObjTriCasterInput {
+	return (timelineObject as TimelineObjTriCasterBase).content?.type === TimelineContentTypeTriCaster.INPUT
+}
+
+export interface TriCasterAudioChannel {
+	isMuted?: boolean
+	/**
+	 * Volume (dB)
+	 * Default: 0
+	 */
+	volume?: number
+}
+
+export interface TimelineObjTriCasterAudioChannel extends TimelineObjTriCasterBase {
+	content: {
+		deviceType: DeviceType.TRICASTER
+		type: TimelineContentTypeTriCaster.AUDIO_CHANNEL
+
+		audioChannel: TriCasterAudioChannel
+	} & TimelineDatastoreReferencesContent
+}
+
+export function isTimelineObjTriCasterAudioChannel(
+	timelineObject: TSRTimelineObjBase
+): timelineObject is TimelineObjTriCasterAudioChannel {
+	return (timelineObject as TimelineObjTriCasterBase).content?.type === TimelineContentTypeTriCaster.AUDIO_CHANNEL
+}
+
+export interface TimelineObjTriCasterMixOutput extends TimelineObjTriCasterBase {
+	content: {
+		deviceType: DeviceType.TRICASTER
+		type: TimelineContentTypeTriCaster.MIX_OUTPUT
+
+		/**
+		 * Any of the named Inputs, Media Players and Buffers ('INPUTn', 'DDRn', 'BFRn') e.g. 'INPUT12' or
+		 * any of the MEs ('Vn') e.g. 'V1' or
+		 * or 'Program', 'Preview', 'program_clean', 'me_program', 'me_preview'
+		 */
+		source: TriCasterMixOutputSource
+	} & TimelineDatastoreReferencesContent
+}
+
+export function isTimelineObjTriCasterMixOutput(
+	timelineObject: TSRTimelineObjBase
+): timelineObject is TimelineObjTriCasterMixOutput {
+	return (timelineObject as TimelineObjTriCasterBase).content?.type === TimelineContentTypeTriCaster.MIX_OUTPUT
+}
+
+export type TriCasterTransitionEffect = 'cut' | 'fade' | number
+
+export interface TriCasterTransition {
+	effect: TriCasterTransitionEffect
+	/** Duration in seconds, applicable to effects other than 'cut' */
+	duration: number
+}
+
+/**
+ * Properties of a layer in effect mode (as opposed to transition mode)
+ * Value ranges in this type adhere to the API and may differ from the GUI
+ */
+export interface TriCasterLayer {
```

Comment:
Logically it means all the properties will be at their defaults. Which properties would you suggest to make required? I don't think it'll work with `input`, for example. The timeline objects support deep merging from multiple layers. I'm not saying this is exactly how we're using it, but consider the following scenario: you have two layers targetting the same M/E. One layer is used for the _DVE_ layout - you'll have four layers with some positioning, but undefined inputs. Then on another layer you will target the same M/E but only provide the inputs for each layer. Now multiply this scenario by every combination of properties that you might want to split into multiple layers.

Reply:
I can see how it can become unmaintainable if we model every variation. I don't understand the combinations well enough to come with concrete suggestions on what to do. My intent was with this as with previous comments on TriCasterMixEffect, was to model the boundarie that seprates the different layer types. E.g. the DVE could conform to a `TriCasterCompositionLayer` interface or some other interface/type that encapsulates a more complex layer.